idunn/api/instant_answer.py

```diff
@@ -267,53 +274,79 @@ async def fetch_pj_response():
     fetch_bragi_osm = asyncio.create_task(bragi_client.autocomplete(query))
     bragi_tripadvisor_response = await bragi_client.autocomplete(query_tripadvisor)
     bragi_tripadvisor_features = result_filter.filter_bragi_features(
         normalized_query, bragi_tripadvisor_response["features"]
     )
 
-    for bragi_tripadvisor_feature in bragi_tripadvisor_features:
-        feature_properties = bragi_tripadvisor_feature["properties"]["geocoding"]
-        if (
-            "poi_types" in feature_properties
-            and feature_properties["poi_types"][0]["id"].split(":")[1] == "subclass_hotel"
-        ):
-            fetch_pj.cancel()
-            fetch_bragi_osm.cancel()
-            place_id = feature_properties["id"]
-            return await run_in_threadpool(
-                get_instant_answer_single_place,
-                query=q,
-                place_id=place_id,
-                lang=lang,
-                type="poi_tripadvisor",
-            )
-
+    # Select datasource instant answer in France
+    if len(intentions) > 0 and pj_source.bbox_is_covered(intentions[0].filter.bbox):
+        for bragi_tripadvisor_feature in bragi_tripadvisor_features:
+            feature_properties = bragi_tripadvisor_feature["properties"]["geocoding"]
+            if (
+                "poi_types" in feature_properties
+                and feature_properties["poi_types"][0]["id"].split(":")[1] == "subclass_hotel"
+            ):
+                fetch_pj.cancel()
+                fetch_bragi_osm.cancel()
+                place_id = feature_properties["id"]
+                return await run_in_threadpool(
+                    get_instant_answer_single_place,
+                    query=q,
+                    place_id=place_id,
+                    lang=lang,
+                    type="poi_tripadvisor",
+                )
+
+            pj_response = result_filter.filter_places(normalized_query, await fetch_pj)
+
+            if pj_response:
+                fetch_bragi_osm.cancel()
+                place_id = pj_response[0].get_id()
+                result = InstantAnswerResult(
+                    places=[pj_response[0].load_place(lang)],
+                    source=pj_response[0].get_source(),
+                    maps_url=maps_urls.get_place_url(place_id),
+                    maps_frame_url=maps_urls.get_place_url(place_id, no_ui=True),
+                )
+                return build_response(result, query=q, lang=lang)
+    # Call tripadvisor datasource instant answer outside France or
+    # without intention location detection
+    else:
+        for bragi_tripadvisor_feature in bragi_tripadvisor_features:
+            feature_properties = bragi_tripadvisor_feature["properties"]["geocoding"]
+            if (
+                "poi_types" in feature_properties
+                and feature_properties["poi_types"][0]["id"].split(":")[0]
+                in AVAILABLE_CLASS_TYPE_TRIPADVISOR
```

Comment:
I feel like there is some code duplication with the french part, maybe we could build a list `allowed_classes_types` which would be either `AVAILABLE_CLASS_TYPE_TRIPADVISOR` or `["class_hotel"]` depending on the location?

Reply:
Indeed there is some code duplication, but I can't easily refactor with interruption workflow in the loop. Probably, need another PR to refactor more largely instant answer